cmd/ceremony/key.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+
+	"github.com/letsencrypt/boulder/pkcs11helpers"
+	"github.com/miekg/pkcs11"
+)
+
+type generateArgs struct {
+	mechanism    []*pkcs11.Mechanism
+	privateAttrs []*pkcs11.Attribute
+	publicAttrs  []*pkcs11.Attribute
+}
+
+func getRandomBytes(ctx pkcs11helpers.PKCtx, session pkcs11.SessionHandle) ([]byte, error) {
+	r, err := ctx.GenerateRandom(session, 4)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+const (
+	rsaModLen = 2048
```

Comment:
Does this imply that even roots are hardcoded to 2048 bits? Our current RSA root is 4096 bits and I suspect we'd want the same if we generated another one. If we expect not to support RSA root generation for now we should document that (and enforce in code).

Reply:
Yeah, we talked OOB about this recently and decided against allowing arbitrary mod lengths, perhaps it'd be better to allow configuring it, but only allow 2048 or 4096.